### PR TITLE
Type comparison

### PR DESCRIPTION
**Pull request by grenki**

### CODE REVIEW

**Comment by zolotov** [comment]
The PR looks strange, rebase it please

File: gen/com/goide/parser/GoParser.java [hunk 1]
```diff
@@ -4388,9 +4388,7 @@ static boolean unary_op(PsiBuilder b, int l) {
   // 4: BINARY(MulExpr)
   // 5: PREFIX(UnaryExpr)
   // 6: ATOM(ConversionExpr)
-  // 7: ATOM(CompositeLit) ATOM(OperandName) POSTFIX(BuiltinCallExpr) POSTFIX(CallExpr)
-  //    POSTFIX(TypeAssertionExpr) BINARY(SelectorExpr) POSTFIX(IndexOrSliceExpr) ATOM(Literal)
-  //    ATOM(FunctionLit)
+  // 7: ATOM(CompositeLit) ATOM(OperandName) POSTFIX(BuiltinCallExpr) POSTFIX(CallExpr) POSTFIX(TypeAssertionExpr) BINARY(SelectorExpr) POSTFIX(IndexOrSliceExpr) ATOM(Literal) ATOM(FunctionLit)
```
Comment (ignatov) [inline gen/com/goide/parser/GoParser.java]: Please regen with the latest Grammar-Kit from the plugin repo.

File: src/com/goide/psi/impl/GoPsiImplUtil.java [hunk 2]
```diff
@@ -737,6 +736,30 @@ public boolean value(@NotNull GoMethodSpec spec) {
   }
 
   @NotNull
+  public static List<GoMethodSpec> getAllMethods(@NotNull GoInterfaceType interfaceType) {
+    List<GoMethodSpec> result = ContainerUtil.newSmartList();
+    getAllMethods(interfaceType, result);
+    return result;
+  }
+
+  private static void getAllMethods(@NotNull GoInterfaceType interfaceType, @NotNull List<GoMethodSpec> result) {
```
Comment (ignatov) [inline src/com/goide/psi/impl/GoPsiImplUtil.java]: Don't use out parameters, looks like C. Much better to create a function which returns List<GoMethodSpec>.

Comment (ignatov) [inline src/com/goide/psi/impl/GoPsiImplUtil.java]: InterfaceType should be nullable.

File: src/com/goide/psi/impl/GoPsiImplUtil.java [hunk 3]
```diff
@@ -1460,6 +1483,14 @@ public static GoExpression getSendExpression(@NotNull GoSendStatement sendStatem
     return getLastExpressionAfter(sendStatement.getExpressionList(), sendStatement.getSendChannel());
   }
 
+  @NotNull
+  public static ChannelDirection getDirection(@NotNull GoChannelType channel) {
+    PsiElement send = channel.getSendChannel();
+    PsiElement chan = channel.getChan();
+    if (send == null || chan == null) return ChannelDirection.BIDIRECTIONAL;
+    return chan.isEquivalentTo(send.getNextSibling()) ? ChannelDirection.SEND : ChannelDirection.RECEIVER;
```
Comment (ignatov) [inline src/com/goide/psi/impl/GoPsiImplUtil.java]: send.getNextSibling could be a space.

File: src/com/goide/psi/impl/GoTypeUtil.java [hunk 4]
```diff
@@ -121,7 +122,8 @@ public static boolean isString(@Nullable GoType type) {
   @NotNull
   private static List<GoType> getExpectedTypesFromGoSendStatement(@NotNull GoExpression expression, @NotNull GoSendStatement statement) {
     GoLeftHandExprList leftHandExprList = statement.getLeftHandExprList();
-    GoExpression channel = ContainerUtil.getFirstItem(leftHandExprList != null ? leftHandExprList.getExpressionList() : statement.getExpressionList());
+    GoExpression channel =
```
Comment (ignatov) [inline src/com/goide/psi/impl/GoTypeUtil.java]: ??

File: src/com/goide/psi/impl/GoTypeUtil.java [hunk 5]
```diff
@@ -231,7 +233,7 @@ public static boolean isString(@Nullable GoType type) {
       }
       return result;
     }
-    
+
```
Comment (ignatov) [inline src/com/goide/psi/impl/GoTypeUtil.java]: The same

File: src/com/goide/psi/impl/GoPsiImplUtil.java [hunk 6]
```diff
@@ -736,6 +736,27 @@ public boolean value(@NotNull GoMethodSpec spec) {
...
+      if (method.getIdentifier() != null) {
+        result.add(method);
+      }
+      else {
+        GoTypeReferenceExpression reference = method.getTypeReferenceExpression();
+        if (reference == null) continue;
+        PsiElement resolve = reference.resolve();
+        if (resolve instanceof GoTypeSpec) {
+          GoInterfaceType innerInterface = ObjectUtils.tryCast(((GoTypeSpec)resolve).getSpecType().getType(), GoInterfaceType.class);
+          if (innerInterface != null) result.addAll(getAllMethods(innerInterface));
```
Comment (zolotov) [inline src/com/goide/psi/impl/GoPsiImplUtil.java]: redundant `!= null` check

**Pull request closed without merging**